docs/proposal/20211207-delegated-storage.md

```diff
@@ -0,0 +1,1186 @@
+    }
+
+    message GetLockedByResponse {
+      string locked_by = 1;
+    }
+
+    message OnLockUnlockRequest {
+      string typeinstance_id = 1;
+      google.protobuf.Any additional_parameters = 2;
+      string locked_by = 3;
+    }
+
+    message OnLockUnlockResponse {}
+
+    // services
+
+    service StorageBackend {
+      // value
+      rpc GetValue(GetValueRequest) returns (GetValueResponse);
+      rpc OnCreate(OnCreateRequest) returns (OnCreateResponse);
+      rpc OnUpdate(OnUpdateRequest) returns (OnUpdateResponse);
+      rpc OnDelete(OnDeleteRequest) returns (OnDeleteResponse);
+
+      // lock
+      rpc GetLockedBy(GetLockedByRequest) returns (GetLockedByResponse);
+      rpc OnLock(OnLockUnlockRequest) returns (OnLockUnlockResponse);
+      rpc OnUnlock(OnLockUnlockRequest) returns (OnLockUnlockResponse);
+    } 
+    ```
+
+    </details>
+
+    An implementation of such service may vary between two use cases:
+
+    1. CRUD operations on output TypeInstance actually manages external resource (e.g. Vault) -> onCreate, onUpdate, and onDelete actually creates, updates and deletes a given resource.
+    1. output TypeInstance represents external resources managed in different way (e.g. via Capact actions - like Helm Runner). IMO we shouldn't move actual Helm release installation to TypeInstance "constructor").
+
+        - The service can also implement watch for external resources (e.g. Kubernetes secrets) and call `createTypeInstances` and `deleteTypeInstances` Hub mutations. We may provide Go framework to speed up such development, similarly as we have with Runner concept.
+
+1. The service could be implemented using one of the following solutions, or other alternatives:
+
+  - [Dapr secrets](https://docs.dapr.io/developing-applications/building-blocks/secrets/secrets-overview/)
+  - [Kubernetes external secrets](https://github.com/external-secrets/kubernetes-external-secrets)
+  - [vault-k8s](https://github.com/hashicorp/vault-k8s)
+  - [db](https://upper.io/v4/getting-started/)
+  - [go-cloud](https://github.com/google/go-cloud)
+  - [stow](https://github.com/graymeta/stow)
+
+## Configuring default storage backends
+
+1. System Administrator configures default backends in Policy.
+
+    ```yaml
+    rules: [...] # rules for Interfaces
+
+    # ...
+
+    default:
+      typeInstance:
+        backend:
+          id: a36ed738-dfe7-45ec-acd1-8e44e8db893b
+          description: "Default Capact PostgreSQL backend"
+    ```
+
+    Default storage backend should have `additionalParameters` empty (`null`) or optional, in order to work properly. When updating Policy, the default backend storage will be validated to see if it meets the criteria.
+
+1. To make it easy to define default backend storages, in Global and Action Policies we introduce additional feature - default TypeInstance injection configuration:
+
+    ```yaml
+    rules: [...] # rules for Interfaces
+
+    common: # properties applied to all rules above
```

Comment:
But then it would be good to define it once and do not repeat such definitions. If we have it defined for `cap.*`, then it's hard guess it is shared along all others entries in rules. I would like to specify "always inject these TIs" and be able to override them one by one.

Reply:
> I would like to specify "always inject these TIs" once and be able to override them one by one.

Totally agree with this, still IMO we should correlate the generic **common** property name with **interfaces**

as we already have a syntax for it maybe just reuse it? as you can see, I still don't know what to put instead of `oneOf` property name, maybe **default**?

Reply:
Sounds fine for me 👍 We can definitely have similar syntax under `common`.